apps/emqx_resource/src/emqx_resource_buffer_worker.erl

```diff
@@ -1020,47 +1034,99 @@ do_handle_async_reply(
         ref => Ref,
         result => Result
     }),
-
+    IsFullBefore = is_inflight_full(InflightTID),
     case Action of
         nack ->
             %% Keep retrying.
-            mark_inflight_as_retriable(InflightTID, Ref),
+            ok = mark_inflight_as_retriable(InflightTID, Ref),
             ?MODULE:block(Pid);
         ack ->
-            do_ack(InflightTID, Ref, Id, Index, PostFn, Pid, QueryOpts)
-    end.
+            do_async_ack(InflightTID, Ref, Id, Index, PostFn, QueryOpts)
+    end,
+    ok = maybe_flush_after_async_reply(IsFullBefore).
```

Comment:
This should only run in the `ack` branch, right?  Otherwise, we will cast a `block` and a `flush` at the same time if the inflight is true?

Reply:
it's fine to cast a `flush` after `block` because the message is discarded.

Reply:
for this one (non-batch), there are not many branches, so I changed it per your suggestion.